src/logic/nontrivial.lean

```diff
@@ -128,3 +133,58 @@ begin
   { exact ⟨x₁, (hf.ne_iff' h).2 hx⟩ },
   { exact ⟨x₂, h⟩ }
 end
+
+namespace tactic
+
+/--
+Given a goal `a = b` or `a ≤ b` in a type `α`, generates an additional hypothesis `nontrivial α`
+(as otherwise `α` is a subsingleton and the goal is trivial).
+-/
+meta def nontriviality_by_elim : tactic unit :=
+do
+  t ← (do `(%%a = %%b) ← target, infer_type a) <|> (do `(%%a ≤ %%b) ← target, infer_type a) <|>
+    fail "Goal is not `_ = _` or `_ ≤ _`",
+  alternative ← to_expr ``(subsingleton_or_nontrivial %%t),
+  tactic.cases alternative,
+  `[{ resetI, try { apply le_of_eq }, apply subsingleton.elim, }] <|>
+    fail format!"Could not prove goal assuming `subsingleton {t}`",
+  reset_instance_cache
+
+/-
+Given a goal `a ≠ b` or `a < b` in a type `α`, tries to generate a `nontrivial α`
+hypothesis from existing hypotheses using `nontrivial_of_ne` and `nontrivial_of_lt`.
+-/
+meta def nontriviality_by_assumption : tactic unit :=
+do
+  t ← (do `(%%a ≠ %%b) ← target, infer_type a) <|> (do `(%%a < %%b) ← target, infer_type a) <|>
+    fail "Goal is not `_ ≠ _` or `_ < _`",
+  n ← get_unused_name "_inst",
+  to_expr ``(nontrivial %%t) >>= assert n,
+  `[solve_by_elim [nontrivial_of_ne, nontrivial_of_lt]],
+  reset_instance_cache
+
+end tactic
+
+namespace tactic.interactive
+
+open tactic
+
+/--
+Given a goal `a = b` or `a ≤ b` in a type `α`, generates an additional hypothesis `nontrivial α`
+(as otherwise `α` is a subsingleton and the goal is trivial).
+
+Alternatively, given a goal `a ≠ b` or `a < b` in a type `α`, tries to generate a `nontrivial α`
+hypothesis from existing hypotheses using `nontrivial_of_ne` and `nontrivial_of_lt`.
+-/
+meta def nontriviality : tactic unit :=
+nontriviality_by_elim <|>
+  nontriviality_by_assumption <|>
```

Comment:
I would swap these.  `nontriviality_by_assumption` can be safely used in definitions, but `nontriviality_by_elim` is noncomputable.